File: ImgProcessing.java
/* 
 * Java port of ImgProcessing.py.
 * Each image processing function should be able to take as arguments
 * scalars, x, y or r,a coordinates, or images as arrays of doubles,
 * and return an image as an array of doubles.
 *
 * Amar Dinsa 2010
 */

import java.awt.image.*;
import java.awt.*;
import java.io.*;
import javax.imageio.ImageIO;

public class ImgProcessing
{

	// Domain of X and Y values (used in scalarToImg)
	private static final int[][] domain = {{-1,1},{-1,1}};

	private static int width, height;

	public ImgProcessing(int w, int h)
	{
		this.width = w;
		this.height = h;
	}

	public double[] add(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = aImg[i] + bImg[i];
		}

		return this.scaleImg(imgData);
	}

	public double[] sub(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = aImg[i] - bImg[i];
		}

		return this.scaleImg(imgData);
	}

	public double[] mul(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = aImg[i] * bImg[i];
		}

		return this.scaleImg(imgData);
	}

	public double[] div(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			double result = 0.0;
			if (bImg[i] != 0) 
				result = (double)aImg[i] / (double)bImg[i];
			imgData[i] = result;
		}

		return this.scaleImg(imgData);
	}

	public double[] mod(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			double result = 0.0;
			if (bImg[i] != 0) 
				result = (double)aImg[i] % (double)bImg[i];
			imgData[i] = result;
		}

		return this.scaleImg(imgData);
	}

	public double[] rnd(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = (double)Math.round(aImg[i]*Math.pow(10,(int)bImg[i])) / (Math.pow(10,(int)bImg[i]));
		}

		return this.scaleImg(imgData);
	}

	public double[] minimum(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = Math.min(aImg[i], bImg[i]);
		}

		return this.scaleImg(imgData);
	}


	public double[] maximum(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = Math.max(aImg[i], bImg[i]);
		}

		return this.scaleImg(imgData);
	}


	public double[] absolute(Object a)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = (double)Math.abs(aImg[i]);
		}

		return this.scaleImg(imgData);
	}

	public double[] expt(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			// Negative number cannot be raised to fractional power
			// Zero cannot be raised to negative power.
			if (!(aImg[i] == 0 && bImg[i] < 0))
			{
				if (aImg[i] < 0)
					bImg[i] = (int)bImg[i];
				imgData[i] = (double)Math.pow(aImg[i],bImg[i]);
			}
			else
				imgData[i] = 0.0;
		}

		return this.scaleImg(imgData);
	}

	public double[] logarithm(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			// Neither argument can be negative and b can't be 1
			if (aImg[i] > 0 && bImg[i] > 0 && bImg[i] != 1 && bImg[i] != 0)
				imgData[i] = (double)Math.log(aImg[i]) /
					     (double)Math.log(bImg[i]);
			else
				imgData[i] = 0.0;

		}

		return this.scaleImg(imgData);
	}

	public double[] sine(Object a)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = Math.sin(aImg[i]);
		}

		return this.scaleImg(imgData);
	}

	public double[] asine(Object a)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			if (aImg[i] < 1 || aImg[i] > 1)
				imgData[i] = 0.0;
			else
				imgData[i] = Math.asin(aImg[i]);
		}

		return this.scaleImg(imgData);
	}

	public double[] cosine(Object a)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = Math.cos(aImg[i]);
		}

		return this.scaleImg(imgData);
	}

	public double[] acosine(Object a)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			if (aImg[i] < -1 || aImg[i] > 1)
				imgData[i] = 0.0;
			else
				imgData[i] = Math.acos(aImg[i]);
		}

		return this.scaleImg(imgData);
	}

	public double[] tangent(Object a)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = Math.tan(aImg[i]);
		}

		return this.scaleImg(imgData);
	}

	public double[] atangent(Object a)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			imgData[i] = Math.atan(aImg[i]);
		}

		return this.scaleImg(imgData);
	}

	public double[] And(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			String aBin = Integer.toBinaryString(Math.abs((int)(aImg[i]*100)));
			String bBin = Integer.toBinaryString(Math.abs((int)(bImg[i]*100)));
			StringBuffer bitSum = new StringBuffer();
			for (int j = 0; j < Math.min(aBin.length(),bBin.length()); j++)
			{
				if (aBin.charAt(j) == bBin.charAt(j))
					bitSum.append('1');
				else
					bitSum.append('0');
			}
			int finalSum = 0;
			if (!(bitSum.toString().equals("")))
				finalSum = Integer.parseInt(bitSum.toString(), 2);
			imgData[i] = finalSum;
		}

		return this.scaleImg(imgData);
	}

	public double[] Or(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			String aBin = Integer.toBinaryString(Math.abs((int)(aImg[i]*100)));
			String bBin = Integer.toBinaryString(Math.abs((int)(bImg[i]*100)));
			StringBuffer bitSum = new StringBuffer();
			for (int j = 0; j < Math.min(aBin.length(),bBin.length()); j++)
			{
				if (aBin.charAt(j) == '1' || bBin.charAt(j) == '1')
					bitSum.append('1');
				else
					bitSum.append('0');
			}
			int finalSum = 0;
			if (!(bitSum.toString().equals("")))
				finalSum = Integer.parseInt(bitSum.toString(), 2);
			imgData[i] = finalSum;
		}

		return this.scaleImg(imgData);
	}

	public double[] Xor(Object a, Object b)
	{
		double imgData[] = new double[this.width*this.height];
		double aImg[] = this.argToImg(a);
		double bImg[] = this.argToImg(b);

		for (int i = 0; i < (this.width*this.height); i++)
		{
			String aBin = Integer.toBinaryString(Math.abs((int)(aImg[i]*100)));
			String bBin = Integer.toBinaryString(Math.abs((int)(bImg[i]*100)));
			StringBuffer bitSum = new StringBuffer();
			for (int j = 0; j < Math.min(aBin.length(),bBin.length()); j++)
			{
				if ((aBin.charAt(j) == '1' && bBin.charAt(j) == '0') || (aBin.charAt(j) == '0' && bBin.charAt(j) == '1'))
					bitSum.append('1');
				else
					bitSum.append('0');
			}
			int finalSum = 0;
			if (!(bitSum.toString().equals("")))
				finalSum = Integer.parseInt(bitSum.toString(), 2);
			imgData[i] = finalSum;
		}

		return this.scaleImg(imgData);
	}

	/**** Utility Functions ****/

	/* Convert object in, which may by a scalar,
	 * x,y or r,a coordinate, or other image,
	 * into image and return it. */
	private static double[] argToImg(Object in)
	{
		double newImg[] = new double[ImgProcessing.width*ImgProcessing.height];

		if (in.getClass().getName().equals("java.lang.String") && Character.isDigit(((String)in).toCharArray()[((String)in).length()-1]))
			newImg = ImgProcessing.scalarToImg(Double.parseDouble((String)in));

		if (in.getClass().getName().equals("java.lang.String") && 
			(((String)in).trim().equals("x") || ((String)in).trim().equals("y") ||
		         ((String)in).trim().equals("r") || ((String)in).trim().equals("a")))
			newImg = ImgProcessing.varToImg((String)in);

		if (in.getClass().getName().toCharArray()[0] == '[')
			newImg = (double[])in;

		return newImg;
	}

	/* Create an image from scalar value scalar */
	private static double[] scalarToImg(double scalar)
	{
		double newImg[] = new double[ImgProcessing.width*ImgProcessing.height];
		for (int i = 0; i < newImg.length; i++)
			newImg[i] = scalar;

		return newImg;
	}

	/* Take "x" or "y" as cartesian coordinates or "r" or "a"
	 * as polar coordinates and return image data with values
	 * set to given coordinate variable (gradient).
	 */
	private static double[] varToImg(String var)
	{
		double newImg[] = new double[ImgProcessing.width*ImgProcessing.height];
		
		double xCur = (double)ImgProcessing.domain[0][0];
		double yCur = (double)ImgProcessing.domain[1][1];
		double xInter = ((double)ImgProcessing.domain[0][1]-ImgProcessing.domain[0][0])/
				(double)ImgProcessing.width;
		double yInter = ((double)ImgProcessing.domain[1][1]-ImgProcessing.domain[1][0])/(double)ImgProcessing.height;

		if (var.equals("x"))
		{
			int index = 0;
			for (int i = 0; i < ImgProcessing.height; i++)
			{
				for (int j = 0; j < ImgProcessing.width; j++)
				{
					newImg[index++] = xCur;
					xCur += xInter;
				}
				xCur = ImgProcessing.domain[0][0];
			}
		}
		else if (var.equals("y"))
		{
			int index = 0;
			for (int i = 0; i < ImgProcessing.height; i++)
			{
				for (int j = 0; j < ImgProcessing.width; j++)
				{
					newImg[index++] = yCur;
				}

				yCur -= yInter;
			}
		}
		else if (var.equals("r"))
		{
			int index = 0;
			for (int i = 0; i < ImgProcessing.height; i++)
			{
				for (int j = 0; j < ImgProcessing.width; j++)
				{
					double dist = Math.sqrt((xCur*xCur)+(yCur*yCur));
					newImg[index++] = dist;
					xCur += xInter;
				}
				yCur -= yInter;
				xCur = ImgProcessing.domain[0][0];
			}
		}
		else if (var.equals("a"))
		{
			int index = 0;
			for (int i = 0; i < ImgProcessing.height; i++)
			{
				for (int j = 0; j < ImgProcessing.width; j++)
				{
					double angle = Math.atan((double)yCur)/(double)xCur;
					newImg[index++] = angle;
					xCur += xInter;
				}
				yCur -= yInter;
				xCur = ImgProcessing.domain[0][0];
			}
		}

		return newImg;

	}

 	/* Return a new image where all the values have
	 * been scaled between 0 and 1.0. */
	private static double[] scaleImg(double[] imgData)
	{
		double min = imgData[0];
		double max = imgData[0];

		for (int i = 1; i < imgData.length; i++)
		{
			if (imgData[i] < min)
				min = imgData[i];
			if (imgData[i] > max)
				max = imgData[i];
		}

		double newImg[] = new double[imgData.length];
		for (int i = 0; i < imgData.length; i++)
		{
			double x = 0.0;
			if (!((max-min) == 0))
				x = (double)(imgData[i] - min) / (max-min);
			newImg[i] = x;
		}

		return newImg;
	}

	public static void main(String[] args)
	{
		int width = 400;
		int height = 400;

		ImgProcessing ip = new ImgProcessing(width, height);
		double image[] = ip.varToImg("r");
		image = ip.scaleImg(image);

		for (int i = 0; i < width*height; i++)
		{
			image[i] = Math.ceil(image[i]*255);
		}

		BufferedImage bi = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
		Graphics2D g = bi.createGraphics();

		g.setBackground(Color.WHITE);
		g.clearRect(0,0, bi.getWidth(), bi.getHeight());
	
		int index = 0;
		for (int i = 0; i < width; i++)
		{
			for (int j = 0; j < height; j++)
			{
				g.setColor(new Color((int)image[index], (int)image[index],(int) image[index]));
				g.drawRect(i,j, 1,1);
				index++;
			}
		}

		try {
			ImageIO.write(bi, "png", new File("image.png"));
		} catch (IOException e)
		{
			System.out.println("Error!\n");
		}
	}
}

File: ImgProcessing.py
# Contains image processing functions used in PicEvolve.
# Each function should be able to take as arguments
# scalars, x,y or r,a coordinates, or other images as lists
# of scalars, and return an image as a list of scalars.
# These scalar lists are converted to pixel color values
# in PicEvolve using GimpGradient.

import math
import random
import time
from PIL import Image, ImageFilter
from GimpGradient import GimpGradient

class ImgProcessing:

	# Domain of X and Y values 
	Domain = [(-1,1),(-1,1)]

	def __init__(self, imgWidth, imgHeight, gradient):

		self.width = imgWidth
		self.height = imgHeight
		self.colorMap = gradient

	def add(self, a, b):

		imgData = []
		aImg, bImg = self.argsToImgs(a, b)

		for i in range(self.width*self.height):
			imgData.append(aImg[i] + bImg[i])

		return self.scaleImg(imgData)

	def sub(self, a, b):

		imgData = []
		aImg, bImg = self.argsToImgs(a, b)

		for i in range(self.width*self.height):
			imgData.append(aImg[i] - bImg[i])

		return self.scaleImg(imgData)
	
	def mul(self, a, b):
		
		imgData = []
		aImg,bImg = self.argsToImgs(a, b)

		for i in range(self.width*self.height):
			imgData.append(aImg[i] * bImg[i])

		return self.scaleImg(imgData)

	def div(self, a, b):

		imgData = []
		aImg, bImg = self.argsToImgs(a, b)

		for i in range(self.width*self.height):
			if not bImg[i] == 0:
				imgData.append(float(aImg[i])/float(bImg[i]))
			else:
				imgData.append(0)

		return self.scaleImg(imgData)

	def mod(self, a, b):

		imgData = []
		aImg, bImg = self.argsToImgs(a, b)

		for i in range(self.width*self.height):
			if not bImg[i] == 0:
				imgData.append(float(aImg[i]) % float(bImg[i]))
			else:
				imgData.append(0)

		return self.scaleImg(imgData)

	def rnd(self, a, b):

		imgData = []
		aImg, bImg = self.argsToImgs(a, b)

		for i in range(self.width*self.height):
			imgData.append(round(aImg[i],int(bImg[i])))

		return self.scaleImg(imgData)

	def minimum(self, a, b):

		imgData = []
		aImg, bImg = self.argsToImgs(a, b)

		for i in range(self.width*self.height):
			imgData.append(min(aImg[i], bImg[i]))

		return self.scaleImg(imgData)

	def maximum(self, a, b):

		imgData = []
		aImg, bImg = self.argsToImgs(a, b)

		for i in range(self.width*self.height):
			imgData.append(max(aImg[i], bImg[i]))

		return self.scaleImg(imgData)

	def absolute(self, a):

		imgData = []
		aImg, bImg = self.argsToImgs(a, None)

		for i in range(self.width*self.height):
			imgData.append(abs(aImg[i]))

		return self.scaleImg(imgData)

	def expt(self, a, b):

		imgData = []

		aImg, bImg = self.argsToImgs(a,b)
		for i in range(self.width*self.height):
			# Negative number cannot be raised to fractional power
			# Zero cannot be raised to negative power.
			if not (aImg[i] == 0 and bImg[i] < 0):
				if aImg[i] < 0:
					bImg[i] = int(bImg[i])
				imgData.append(pow(aImg[i], bImg[i]))
			else:
				imgData.append(0)

		return self.scaleImg(imgData)

	def logarithm(self, a, b):

		imgData = []

		aImg, bImg = self.argsToImgs(a,b)
		for i in range(self.width*self.height):
			# Neither argument can be negative and b can't be 1
			if aImg[i] > 0 and bImg[i] > 0 and not bImg[i] == 1:
				imgData.append(math.log(aImg[i], bImg[i]))
			else:
				imgData.append(0)

		return self.scaleImg(imgData)

	def sine(self, a):

		imgData = []

		aImg, bImg = self.argsToImgs(a, None)
		for i in range(self.width*self.height):
			imgData.append(math.sin(aImg[i]))

		return self.scaleImg(imgData)

	def asine(self, a):

		imgData = []
		aImg, bImg = self.argsToImgs(a, None)

		for i in range(self.width*self.height):
			if aImg[i] < -1 or aImg[i] > 1:
				imgData.append(0)
			else:
				imgData.append(math.asin(aImg[i]))

		return self.scaleImg(imgData)

	def cosine(self, a):

		imgData = []

		aImg, bImg = self.argsToImgs(a, None)
		for i in range(self.width*self.height):
			imgData.append(math.cos(aImg[i]))

		return self.scaleImg(imgData)

	def acosine(self, a):

		imgData =[]

		aImg,bImg = self.argsToImgs(a, None)
		for i in range(self.width*self.height):
			if aImg[i] < -1 or aImg[i] > 1:
				imgData.append(0)
			else:
				imgData.append(math.acos(aImg[i]))

		return self.scaleImg(imgData)

	def tangent(self, a):

		imgData = []
		aImg,bImg = self.argsToImgs(a, None)
		for i in range(self.width*self.height):
			imgData.append(math.tan(aImg[i]))

		return self.scaleImg(imgData)

	def atangent(self, a):

		imgData = []
		aImg,bImg = self.argsToImgs(a, None)
		for i in range(self.width*self.height):
			imgData.append(math.atan(aImg[i]))

		return self.scaleImg(imgData)

	def And(self, a, b):

		imgData = []

		aImg, bImg = self.argsToImgs(a,b)
		for i in range(self.width*self.height):
			aBin = bin(int(aImg[i]*100))[2:]
			bBin = bin(int(bImg[i]*100))[2:]
			bitSum = ""
			for i in range(min(len(aBin),len(bBin))):
				if aBin[i] == bBin[i]:
					bitSum = bitSum + "1"
				else:
					bitSum = bitSum + "0"
			finalSum = 0
			if not bitSum == "":
				finalSum = eval("0b"+ bitSum)
			imgData.append(finalSum)

		return self.scaleImg(imgData)


	def Or(self, a, b):

		imgData = []

		aImg, bImg = self.argsToImgs(a,b)
		for i in range(self.width*self.height):
			aBin = bin(int(aImg[i]*100))[2:]
			bBin = bin(int(bImg[i]*100))[2:]
			bitSum = ""
			for i in range(min(len(aBin),len(bBin))):
				if aBin[i] == "1" or bBin[i] == "1":
					bitSum = bitSum + "1"
				else:
					bitSum = bitSum + "0"
			finalSum = 0
			if not bitSum == "":
				finalSum = eval("0b"+ bitSum)
			imgData.append(finalSum)

		return self.scaleImg(imgData)

	def Xor(self, a, b):
		imgData = []

		aImg, bImg = self.argsToImgs(a,b)
		for i in range(self.width*self.height):
			aBin = bin(int(aImg[i]*100))[2:]
			bBin = bin(int(bImg[i]*100))[2:]
			bitSum = ""
			for i in range(min(len(aBin),len(bBin))):
				if (aBin[i] == "1" and bBin[i] == "0") or (aBin[i] == "0" and bBin[i] == "1"):
					bitSum = bitSum + "1"
				else:
					bitSum = bitSum + "0"
			finalSum = 0
			if not bitSum == "":
				finalSum = eval("0b"+ bitSum)
			imgData.append(finalSum)

		return self.scaleImg(imgData)

	def noise(self, a, b):
		"""Perlin noise generator with a passed as seed and b
		   as scale."""

		imgData = []

		# First we need to convert a and b into scalars.
		a = self.argToScalar(a)
		b = self.argToScalar(b)
		a = a*10
		b = b*10

		# Now we can seed the PRN generator with argument
		from perlin import Noise2d
		n = Noise2d(a)
		scale = b 
		for x in range(self.width):
			for y in range (self.height):
				val = 0
				if not b == 0.0:
					val = int((n.noise(x/b,y/b)+1)*128)
				imgData.append(val)

		random.seed(time.time())
		return self.scaleImg(imgData)

	def blur(self, a, b):
		"""Blur image a b * 10 times."""

		aImg, bImg = self.argsToImgs(a, b)
		bNum = self.imgToScalar(bImg)

		img = Image.new("RGB",(self.width, self.height))
		pix = img.load()

		count = 0
		for y in range(self.height):
			for x in range(self.width):
				pix[x, y] = self.colorMap.colorRGB(aImg[count])
				count += 1

		for i in range(int(bNum*10)):
			img = img.filter(ImageFilter.BLUR)

		imgData = img.getdata()
		newImg = []
		for r,g,b in imgData:
			newImg.append((r+g+b)/3.0)

		return self.scaleImg(newImg)

	def edgeEnhance(self, a, b):

		aImg, bImg = self.argsToImgs(a, b)
		bNum = self.imgToScalar(bImg)

		img = Image.new("RGB",(self.width, self.height))
		pix = img.load()

		count = 0
		for y in range(self.height):
			for x in range(self.width):
				pix[x, y] = self.colorMap.colorRGB(aImg[count])
				count += 1

		for i in range(int(bNum*10)):
			img = img.filter(ImageFilter.EDGE_ENHANCE)

		imgData = img.getdata()
		newImg = []
		for r,g,b in imgData:
			newImg.append((r+g+b)/3.0)

		return self.scaleImg(newImg)

	def emboss(self, a, b):

		aImg, bImg = self.argsToImgs(a, b)
		bNum = self.imgToScalar(bImg)

		img = Image.new("RGB",(self.width, self.height))
		pix = img.load()

		count = 0
		for y in range(self.height):
			for x in range(self.width):
				pix[x, y] = self.colorMap.colorRGB(aImg[count])
				count += 1

		for i in range(int(bNum*10)):
			img = img.filter(ImageFilter.EMBOSS)

		imgData = img.getdata()
		newImg = []
		for r,g,b in imgData:
			newImg.append((r+g+b)/3.0)

		return self.scaleImg(newImg)

	def mandel(self, a, b):
		"""Mandelbrot fractal with a and b being
		   real and imaginary components."""
		
		aImg, bImg = self.argsToImgs(a, b)
		aNum = self.imgToScalar(aImg)
		bNum = self.imgToScalar(bImg)

		botLeft = complex(aNum,bNum)
		topRight = complex(botLeft.real+3.0,botLeft.imag+3.0)
		xStep = (topRight.real-botLeft.real)/float(self.width)
		yStep = (topRight.imag-botLeft.imag)/float(self.height)

		rawImg = []
		xCur = botLeft.real
		yCur = botLeft.imag
		for y in range(self.height):
			for x in range(self.width):

				color = (0,0,0)
				c = complex(xCur, yCur)
				z = c
				for i in range(50):
					z = z*z + c
					if (z.real*z.real + z.imag*z.imag) > 4:
						color = (i,i,i)
						break
				xCur += xStep
				rawImg.append(color[0])

			yCur += yStep
			xCur = botLeft.real

		return self.scaleImg(rawImg)

	def lerp(self, a, b, c):
		"""Do a linear interpolation with a and b
		   using c as the interpolation factor."""
	
		aImg, bImg = self.argsToImgs(a, b)
		cImg, dummy = self.argsToImgs(c, None)
		cNum = self.imgToScalar(cImg)

		imgData = []
		for i in range(self.width*self.height):
			l = cNum*aImg[i] + (1.0 - cNum)*bImg[i]
			imgData.append(l)

		return self.scaleImg(imgData)


#############################################################################
#                             Uility functions                              #
#############################################################################

	def imgToScalar(self, img):
		"""Return the average value of scalars in img."""
	
		imgSum = 0
		for pixel in img:
			imgSum += pixel
	
		if len(img) == 0:
			return 0
		else:
			return float(imgSum) / float(len(img))

	def varToImg(self, a):
		"""Take "x" or "y" as cartesian coordinates or "r" or "a"
		   as polar coordinates and return image data with values
		   set to given coordinate variable (gradient)."""

		imgData = []
	
		xCur = self.Domain[0][0]
		yCur = self.Domain[1][1]
		xInter = float(self.Domain[0][1]-self.Domain[0][0]) / float(self.width)
		yInter = float(self.Domain[1][1]-self.Domain[1][0]) / float(self.height)
	
		if a == 'x':
			for i in range(self.height):
				for j in range(self.width):
					imgData.append(xCur)
					xCur += xInter
				xCur = self.Domain[0][0]

		elif a == 'y':
			for i in range(self.height):
				for j in range(self.width):
					imgData.append(yCur)
				yCur -= yInter
		elif a == 'r':
			for i in range(self.height):
				for j in range(self.width):
					dist = math.sqrt((xCur*xCur)+(yCur*yCur))
					imgData.append(dist)
					xCur += xInter
				yCur -= yInter
				xCur = self.Domain[0][0]
		elif a == 'a':
			for i in range(self.height):
				for j in range(self.width):
					angle = math.atan(float(yCur)/float(xCur))
					imgData.append(angle)
					xCur += xInter
				yCur -= yInter
				xCur = self.Domain[0][0]

		return imgData

	def scaleImg(self, imgData):
		"""Return a new image where all the values in
		   imgData have been scaled between 0 and 1.0."""
	
		imin = min(imgData)
		imax = max(imgData)

		newImg = []
		for pixel in imgData:
			x = 0
			if not (imax-imin) == 0:
				x = (pixel - imin)/float((imax - imin))
			newImg.append(x)

		return newImg
		

	def argToScalar(self, a):
		"""Convert the given argument, which can be a float,
		   an x,y or r,a coordinate, or an image, into a scalar."""

		if type(a) == list:
			a = self.imgToScalar(a)
		if type(a) == str and a[-1].isdigit():
			a = float(a)
		if type(a) == str and (a == 'x' or a =='y' or a == 'r' or a == 'a'):
			if a == 'x':
				a = 0
			elif a == 'y':
				a = 1
			elif a == 'r':
				a = 2
			elif a == 'a':
				a = 3

		return a

	def scalarToImg(self, a):
		"""Create an image from scalar value a."""
		
		imgData = []

		for i in range(self.width*self.height):
			imgData.append(a)

		return imgData

	def argsToImgs(self, a, b):
		"""Return tuple of images created from a and b,
		   which can be scalars, x,y or r,a coordinates, or images."""

		aImg = []
		bImg = []

		if type(a) == list:
			aImg = a
		if type(b) == list:
			bImg = b

		if type(a) == str and a[-1].isdigit():
			aImg = self.scalarToImg(float(a))
		if type(b) == str and b[-1].isdigit():
			bImg = self.scalarToImg(float(b))
	
		if type(a) == str and (a == 'x' or a =='y' or a == 'r' or a == 'a'):
			aImg = self.varToImg(a)
		if type(b) == str and (b == 'x' or b == 'y' or b == 'r' or b == 'a'):
			bImg = self.varToImg(b)

		return aImg, bImg



if __name__ == "__main__":

	img = Image.new("RGB",(300,300),(255,255,255))
	g = GimpGradient("./gradients/Incandescent.ggr")
